### PR TITLE
Move service reject ACLs to a cluster port group

### PR DESCRIPTION
**Pull request by vishnoianil**
Adding an ACL-per-service-per-node switch doesn't scale. OVN team instead
suggested adding the ACL-per-service to a Port Group which all logical
switch ports are members of.

Note that this does now deny multicast traffic across the management port,
but I can't think of a good reason why that should be allowed or even work
today anyway.

Reported-at: https://bugzilla.redhat.com/show_bug.cgi?id=1855408
Signed-off-by: Anil Vishnoi <vishnoianil@gmail.com>
Signed-off-by: Dan Williams <dcbw@redhat.com>

### CODE REVIEW

**Comment by danwinship** [comment]
Would it be possible to just have a low-priority rule that rejects any unmatched packets to the service CIDR? Or would that not work because we don't otherwise have ACLs accepting valid service IPs?

**Comment by dcbw** [comment]
@vishnoianil is this one still WIP?

**Comment by dcbw** [comment]
@vishnoianil also seems like there's a problem in my original patches, perhaps? CI is failing with:

`{"msg":"FAILED [sig-network] Services should be rejected when no endpoints exist","total":-1,"completed":0,"skipped":3,"failed":1,"failures":["[sig-network] Services should be rejected when no endpoints exist"]}`

**Comment by vishnoianil** [comment]
@dcbw @trozet PTAL, Thanks

**Comment by vishnoianil** [comment]
> Would it be possible to just have a low-priority rule that rejects any unmatched packets to the service CIDR? Or would that not work because we don't otherwise have ACLs accepting valid service IPs?

@danwinship Sorry missed this comment. I think it's one possibility, as far as we can explicitly install a wildcard rule to reject all traffic by default and only allow ARP,IPv6 ND packets. Is there any other traffic that we should be allowing ? I am bit hesitant of default deny policy, but if we know what all traffic we need to allow, i think we can have workable solution. 
 
I think probably the better solution would be to improve the Load Balancer to reject the packet if there are no end-points. @trozet opened an RFE to OVN for that (https://bugzilla.redhat.com/show_bug.cgi?id=1883957).

**Comment by trozet** [comment]
"Note that this does now deny multicast traffic across the management port,
but I can't think of a good reason why that should be allowed or even work
today anyway."

won't we need multicast to work between host network pods and k8s networked pods? That would go through mp0.

File: go-controller/pkg/ovn/loadbalancer.go [hunk 1]
```diff
@@ -230,39 +229,38 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb string, sourceIP string, s
...
+				lb, aclUUID, aclName, stderr, err)
+			return "", err
 		}
+
 		ovn.setServiceACLToLB(lb, vip, aclUUID)
+
+		// If reject ACL exist, ensures that the _uuid is removed from logical_switch acls list.
+		// This step is required to ensure the clean-up when ovn upgrades from logical_switch acls
+		// to port_group based acls.
+		ovn.removeACLFromNodeSwitches(switches, aclUUID)
```
Comment (trozet) [inline go-controller/pkg/ovn/loadbalancer.go:244]: this should be handled in the sync method I think.

File: go-controller/pkg/ovn/loadbalancer.go [hunk 2]
```diff
@@ -230,39 +229,38 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb string, sourceIP string, s
...
-		// backends are re-added.
-		ovn.setServiceACLToLB(lb, vip, aclUUID)
+		klog.Errorf("Failed to add LB %s ACL %s %q to cluster port group: %q (%v)",
+			lb, aclUUID, aclName, stderr, err)
+		return "", err
 	}
+
+	// Associate ACL UUID with load balancer and ip+port so we can remove this ACL if
+	// backends are re-added.
+	ovn.setServiceACLToLB(lb, vip, aclUUID)
```
Comment (trozet) [inline go-controller/pkg/ovn/loadbalancer.go:262]: this needs to go back into the else statement. We shouldn't be adding an ACL to the cache if the OVN ACL was not created.

File: go-controller/pkg/ovn/loadbalancer.go [hunk 3]
```diff
@@ -277,50 +275,58 @@ func (ovn *Controller) deleteLoadBalancerRejectACL(lb, vip string) {
...
 		return
-	} else {
-		ovn.removeACLFromNodeSwitches(lb, aclUUID)
 	}
+	ovn.removeACLFromPortGroup(lb, aclUUID)
 	ovn.removeServiceACL(lb, vip)
 }
 
-func (ovn *Controller) removeStaleRejectACL(lb, ip string, port int32) {
+func (ovn *Controller) removeStaleRejectACL(lb, ip string, port int32) (string, error) {
```
Comment (trozet) [inline go-controller/pkg/ovn/loadbalancer.go:292]: this function no longer removes a stale reject ACL, it just finds an acl. Please rename it

File: go-controller/pkg/ovn/loadbalancer.go [hunk 4]
```diff
@@ -277,50 +275,58 @@ func (ovn *Controller) deleteLoadBalancerRejectACL(lb, vip string) {
...
 		} else {
 			klog.Infof("ACL: %s, removed from switches: %s", aclUUID, switches)
 		}
 	}
 }
+
+func (ovn *Controller) removeACLFromPortGroup(lb, aclUUID string) {
+	_, stderr, err := util.RunOVNNbctl("--", "--if-exists", "remove", "port_group", ovn.clusterPortGroupUUID, "acls", aclUUID)
+	if err != nil {
+		klog.Errorf("Failed to remove reject ACL from LB %s: stderr: %q, error: %v", lb, stderr, err)
```
Comment (trozet) [inline go-controller/pkg/ovn/loadbalancer.go:328]: list the ACL UUID

File: go-controller/pkg/ovn/master.go [hunk 5]
```diff
@@ -262,6 +262,13 @@ func (oc *Controller) SetupMaster(masterNodeName string) error {
 		klog.Info("SCTP support detected in OVN")
 	}
 
+	// Create a cluster-wide port group that all logical switch ports are part of
+	oc.clusterPortGroupUUID, err = createPortGroup(clusterPortGroupName, clusterPortGroupName)
```
Comment (trozet) [inline go-controller/pkg/ovn/master.go:266]: we only need a cluster port group with the mp0 ports in it. That will be enough to get the ACLs on each switch. It looks like you are already doing that? So the ocmment on 265 is not accurate?

File: go-controller/pkg/ovn/master.go [hunk 6]
```diff
@@ -420,11 +425,34 @@ func (oc *Controller) syncNodeManagementPort(node *kapi.Node, hostSubnets []*net
...
+		"--", "--may-exist", "lsp-add", node.Name, portName,
+		"--", "lsp-set-addresses", portName, addresses)
 	if err != nil {
-		klog.Errorf("Failed to add logical port to switch, stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+		klog.Errorf("Failed to add logical port to switch, stdout: %q, stderr: %q, error: %v",
+			stdout, stderr, err)
+		return err
+	}
+
+	// UUID must be retrieved separately from the lsp-add transaction since
```
Comment (trozet) [inline go-controller/pkg/ovn/master.go:438]: I get you are adding this as RunOVNNbctl cause you want to backport this downstream, but we should really be using go-ovn bindings here. I guess this is OK for now and we will clean it up after.

File: go-controller/pkg/ovn/ovn.go [hunk 7]
```diff
@@ -318,6 +322,10 @@ func (oc *Controller) Run(wg *sync.WaitGroup) error {
 
 	klog.Infof("Completing all the Watchers took %v", time.Since(start))
 
+	// Remove old multicastDefaultDeny port group now that all ports
```
Comment (trozet) [inline go-controller/pkg/ovn/ovn.go:325]: why dont we just use one port group with mp0 ports in it for the reject acls, and leave multicast alone?

File: go-controller/pkg/ovn/service.go [hunk 8]
```diff
@@ -168,7 +168,7 @@ func (ovn *Controller) syncServices(services []interface{}) {
 					for lb, hasEps := range svcCacheEntry {
 						if hasEps {
 							klog.Infof("Service Sync: Removing OVN stale reject ACL: %s", name)
-							ovn.removeACLFromNodeSwitches(lb, uuid)
```
Comment (trozet) [inline go-controller/pkg/ovn/service.go:171]: you need to add code into syncServices to handle removing reject acls from switches then let the add handler  add them back as port group for upgrade.

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/master.go (hunk 5 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/master.go:266]: This patch do add all the logical ports to the port-group. I didn't find any document that points out that adding management port will be enough. As per our email communication with ovn team, the idea was to add all the logical ports to the port group. I recently learned it from Ilya that it should work if you only add management port. I assume that it works, this will reduce the memory usage further. So will remove the code that adds ports to port-group. 
On a side note: Does that mean, if you add management port to the port-group, there is no point in adding any other logical port to that group, because it will anyways apply on the logical switch?

File: go-controller/pkg/ovn/master.go (hunk 6 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/master.go:438]: okay, lets clean that up through new PR.

File: go-controller/pkg/ovn/ovn.go (hunk 7 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/ovn.go:325]: If we deny the multicast traffic on management port, both the port-group will be same right? so do we need two port-group in that case?

File: go-controller/pkg/ovn/service.go (hunk 8 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/service.go:171]: please see my comment above regarding adding it to syncServices.

File: go-controller/pkg/ovn/master.go (hunk 5 above)
Comment (trozet) [reply go-controller/pkg/ovn/master.go:266]: right that's my understanding

File: go-controller/pkg/ovn/loadbalancer.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/ovn/loadbalancer.go:262]: oh I missed the return if theres an error. This is fine.

File: go-controller/pkg/ovn/loadbalancer.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/ovn/loadbalancer.go:244]: yeah this is a bit tricky.  I guess this is OK since it's only when cache is invalid.

File: go-controller/pkg/ovn/service.go (hunk 8 above)
Comment (trozet) [reply go-controller/pkg/ovn/service.go:171]: yeah thats fine. I think we still need ovn.removeACLFromNodeSwitches(lb, uuid) here. In case we upgrade from a version that was using ACL on node switches and the ACLs have gone stale.

File: go-controller/pkg/ovn/ovn.go [hunk 9]
```diff
@@ -318,6 +322,10 @@ func (oc *Controller) Run(wg *sync.WaitGroup) error {
...
+	// have been added to the clusterPortGroup by WatchPods()
+	deletePortGroup("mcastPortGroupDeny")
```
Comment (trozet) [inline go-controller/pkg/ovn/ovn.go:327]: Move this into createDefaultDenyMulticastPolicy()

File: go-controller/pkg/ovn/ovn.go (hunk 7 above)
Comment (trozet) [reply go-controller/pkg/ovn/ovn.go:325]: the problem becomes on upgrade the port group now has a different name, so we have to delete the old one, and then wait for all the pod events to add their ports to the new one. But I guess per our discussion there's no other good option if we want to rename the current port group.

File: go-controller/pkg/ovn/master.go (hunk 5 above)
Comment (dcbw) [reply go-controller/pkg/ovn/master.go:266]: Please triple-confirm this understanding of ACLs first though...

**Comment by dcbw** [comment]
> won't we need multicast to work between host network pods and k8s networked pods? That would go through mp0.

Why would we need that? We don't allow that with openshfit-sdn right now. In the future perhaps we should make this work.

But our multicast is per-namespace, and we have no way to distinguish traffic coming from the management port on a per-namespace basis, because all hostnetwork pods share the same PodIP (eg nodeIP).

File: go-controller/pkg/ovn/master.go (hunk 5 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/master.go:266]: @dcbw Yes, i had a discussion over irc and following bugs are result of that discussion as well
https://bugzilla.redhat.com/show_bug.cgi?id=1885670
https://bugzilla.redhat.com/show_bug.cgi?id=1885668

File: go-controller/pkg/ovn/ovn.go (hunk 7 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/ovn.go:325]: thanks for summarizing it tim.

File: go-controller/pkg/ovn/ovn.go (hunk 9 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/ovn.go:327]: Fixed.

File: go-controller/pkg/ovn/service.go (hunk 8 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/service.go:171]: yeah makes sense. Fixed it.

**Comment by trozet** [comment]
> > won't we need multicast to work between host network pods and k8s networked pods? That would go through mp0.
> 
> Why would we need that? We don't allow that with openshfit-sdn right now. In the future perhaps we should make this work.
> 
> But our multicast is per-namespace, and we have no way to distinguish traffic coming from the management port on a per-namespace basis, because all hostnetwork pods share the same PodIP (eg nodeIP).

So some things learned from our slack conversation:
1. Multicast traffic by default is denied across an ovn-kubernetes cluster (default deny mcast port group)
2. Multicast is enabled on a per namespace basis (via an annotation, adds ports to an allow mcast port group)

Some outstanding questions. @dceara maybe you can comment here? @girishmg FYI
1. As I understand the reason for blocking mcast was because we treated it like broadcast in the past. Now that we have IGMP snooping supported, would it be reasonable to enable mcast by default. (but still keep a config knob to default block it for downstream parity)? The reason I bring this is up is from my basic understanding of k8s the default traffic is "anything can talk to anything" and nothing should be blocked (hence the network policy behavior).
2. Why do we have 2 port groups (1 for deny and 1 for allow) with the current method. Why not just remove pods in annotated namespaces from the original deny port group?
3. Is there any documentation on how this works in our repo? If not, can we please add some?

If the answer to number 2 above is, "we only really need 1 port group". Then I would suggest we use that 1 port group for managing multicast, and then create another one with just management ports in it for handling service reject ACL.

**Comment by dceara** [comment]
> > > won't we need multicast to work between host network pods and k8s networked pods? That would go through mp0.
> > 
> > 
> > Why would we need that? We don't allow that with openshfit-sdn right now. In the future perhaps we should make this work.
> > But our multicast is per-namespace, and we have no way to distinguish traffic coming from the management port on a per-namespace basis, because all hostnetwork pods share the same PodIP (eg nodeIP).
> 
> So some things learned from our slack conversation:
> 
>     1. Multicast traffic by default is denied across an ovn-kubernetes cluster (default deny mcast port group)
> 
>     2. Multicast is enabled on a per namespace basis (via an annotation, adds ports to an allow mcast port group)
> 
> 
> Some outstanding questions. @dceara maybe you can comment here? @girishmg FYI
> 
>     1. As I understand the reason for blocking mcast was because we treated it like broadcast in the past. Now that we have IGMP snooping supported, would it be reasonable to enable mcast by default. (but still keep a config knob to default block it for downstream parity)? The reason I bring this is up is from my basic understanding of k8s the default traffic is "anything can talk to anything" and nothing should be blocked (hence the network policy behavior).
> 

I'm not sure about this, I'll let others comment. The default behavior is to drop multicast but it can definitely be changed if needed.

>     2. Why do we have 2 port groups (1 for deny and 1 for allow) with the current method. Why not just remove pods in annotated namespaces from the original deny port group?
> 

We need 2 port groups. When an ACL is applied to a port group that means that it's applied to all logical switches that have ports in the port group. If two pods P1 and P2 from different namespaces NS1 and NS2 reside on the same node N and multicast is enabled only in NS2, removing P2 from the default deny mcast port group is not enough because the deny multicast ACL will still match on all multicast packets on node N.

The alternative would be to use allow ACLs that match on inport/outport but that would result way more flows than now.

>     3. Is there any documentation on how this works in our repo? If not, can we please add some?

There's WIP from network policy in general:
https://github.com/ovn-org/ovn-kubernetes/pull/1182

> 
> 
> If the answer to number 2 above is, "we only really need 1 port group". Then I would suggest we use that 1 port group for managing multicast, and then create another one with just management ports in it for handling service reject ACL.

I don't see a problem with having a single port group for multicast deny and service reject ACLs. The requirement for both features is that the port group contains at least one port from all nodes. So why not have a single PG?

There seems to be some confusion in general about how Port_Group ACLs work so I'm leaving this here for reference:
```
      Access control rules that apply to the port group. Applying an ACL
      to a port group has the same effect as applying the ACL to all logical
      lswitches that the ports of the port group belong to.
```
https://github.com/ovn-org/ovn/blob/46b23163a54abb8cef6ae12585985045735aff45/ovn-nb.xml#L1470

**Comment by trozet** [comment]
> > 2. Why do we have 2 port groups (1 for deny and 1 for allow) with the current method. Why not just remove pods in annotated namespaces from the original deny port group?
> > ```
> 
> We need 2 port groups. When an ACL is applied to a port group that means that it's applied to all logical switches that have ports in the port group. If two pods P1 and P2 from different namespaces NS1 and NS2 reside on the same node N and multicast is enabled only in NS2, removing P2 from the default deny mcast port group is not enough because the deny multicast ACL will still match on all multicast packets on node N.

@dceara but when I look in the repo I see:
```
go-controller/pkg/ovn/master_test.go:		"ovn-nbctl --timeout=15 --id=@acl create acl priority=1011 direction=from-lport match=\"inport == @mcastPortGroupDeny && ip4.mcast\" action=drop external-ids:default-deny-policy-type=Egress -- add port_group  acls @acl",
go-controller/pkg/ovn/master_test.go:		"ovn-nbctl --timeout=15 --data=bare --no-heading --columns=_uuid find ACL match=\"outport == @mcastPortGroupDeny && ip4.mcast\" action=drop external-ids:default-deny-policy-type=Ingress",
```
^we match on inport and output there, so since the port would not be in the deny group anymore it wouldn't be blocked?

**Comment by dceara** [comment]
> > > 1. Why do we have 2 port groups (1 for deny and 1 for allow) with the current method. Why not just remove pods in annotated namespaces from the original deny port group?
> > 
> > 
> > We need 2 port groups. When an ACL is applied to a port group that means that it's applied to all logical switches that have ports in the port group. If two pods P1 and P2 from different namespaces NS1 and NS2 reside on the same node N and multicast is enabled only in NS2, removing P2 from the default deny mcast port group is not enough because the deny multicast ACL will still match on all multicast packets on node N.
> 
> @dceara but when I look in the repo I see:
> 
> ```
> go-controller/pkg/ovn/master_test.go:		"ovn-nbctl --timeout=15 --id=@acl create acl priority=1011 direction=from-lport match=\"inport == @mcastPortGroupDeny && ip4.mcast\" action=drop external-ids:default-deny-policy-type=Egress -- add port_group  acls @acl",
> go-controller/pkg/ovn/master_test.go:		"ovn-nbctl --timeout=15 --data=bare --no-heading --columns=_uuid find ACL match=\"outport == @mcastPortGroupDeny && ip4.mcast\" action=drop external-ids:default-deny-policy-type=Ingress",
> ```
> 
> ^we match on inport and output there, so since the port would not be in the deny group anymore it wouldn't be blocked?

Right, my bad, for the default deny we could just remove the port from the port group.

However the current multicast support allows traffic only between pods in the same namespace (if multicast is enabled). That means that the allow ACLs need to match on the per NS port group/address set. So we still need the default port group and a per namespace port group.

**Comment by trozet** [comment]
OK so now that I also learned allowing mcast does not mean allow mcast from the pod to all other allowed pods (it means only allow from this pod to other pods in this namespace), we still need the mcast allow policy.  However, we do not need to put all the ports in the mcast deny policy.

@vishnoianil I suggest we create a first commit that:
1.  rename mcast port group to allnodeswitch port group or something
2.  do not add all ports into this port group, just mp0s
3.  modify the default deny mcast ACL to not match on inport or outport

Create a second commit that then moves service reject ACL to use this new port group.

**Comment by danwinship** [comment]
Just to confirm, you are _not_ going to enable multicast between the pod network and the host network, right? Because that would be a security hole (potentially allowing pods to interfere with node-to-node multicast traffic).

**Comment by vishnoianil** [comment]
> > > won't we need multicast to work between host network pods and k8s networked pods? That would go through mp0.
> > 
> > 
> > Why would we need that? We don't allow that with openshfit-sdn right now. In the future perhaps we should make this work.
> > But our multicast is per-namespace, and we have no way to distinguish traffic coming from the management port on a per-namespace basis, because all hostnetwork pods share the same PodIP (eg nodeIP).
> 
> So some things learned from our slack conversation:
> 
> 1. Multicast traffic by default is denied across an ovn-kubernetes cluster (default deny mcast port group)
> 2. Multicast is enabled on a per namespace basis (via an annotation, adds ports to an allow mcast port group)
> 
> Some outstanding questions. @dceara maybe you can comment here? @girishmg FYI
> 
> 1. As I understand the reason for blocking mcast was because we treated it like broadcast in the past. Now that we have IGMP snooping supported, would it be reasonable to enable mcast by default. (but still keep a config knob to default block it for downstream parity)? The reason I bring this is up is from my basic understanding of k8s the default traffic is "anything can talk to anything" and nothing should be blocked (hence the network policy behavior).
I won't put multicast traffic in "anything can talk to anything" category, it's still a broadcast traffic, but controlled broadcast, so in my opinion, we should be careful about allowing any broadcast traffic (except ARP/ND) in general.
> 2. Why do we have 2 port groups (1 for deny and 1 for allow) with the current method. Why not just remove pods in annotated namespaces from the original deny port group?
> 3. Is there any documentation on how this works in our repo? If not, can we please add some?
> 
> If the answer to number 2 above is, "we only really need 1 port group". Then I would suggest we use that 1 port group for managing multicast, and then create another one with just management ports in it for handling service reject ACL.

**Comment by vishnoianil** [comment]
> OK so now that I also learned allowing mcast does not mean allow mcast from the pod to all other allowed pods (it means only allow from this pod to other pods in this namespace), we still need the mcast allow policy. However, we do not need to put all the ports in the mcast deny policy.
> 
> @vishnoianil I suggest we create a first commit that:
> 
> 1. rename mcast port group to allnodeswitch port group or something
> 2. do not add all ports into this port group, just mp0s
> 3. modify the default deny mcast ACL to not match on inport or outport
1 & 2, i already have locally and testing, will add 3 as well.
> 
> Create a second commit that then moves service reject ACL to use this new port group.
Yeah, separating these commit makes sense. Will do that.

**Comment by vishnoianil** [comment]
> > > > won't we need multicast to work between host network pods and k8s networked pods? That would go through mp0.
> > > 
> > > 
> > > Why would we need that? We don't allow that with openshfit-sdn right now. In the future perhaps we should make this work.
> > > But our multicast is per-namespace, and we have no way to distinguish traffic coming from the management port on a per-namespace basis, because all hostnetwork pods share the same PodIP (eg nodeIP).
> > 
> > 
> > So some things learned from our slack conversation:
> > ```
> > 1. Multicast traffic by default is denied across an ovn-kubernetes cluster (default deny mcast port group)
> > 
> > 2. Multicast is enabled on a per namespace basis (via an annotation, adds ports to an allow mcast port group)
> > ```
> > 
> > 
> > Some outstanding questions. @dceara maybe you can comment here? @girishmg FYI
> > ```
> > 1. As I understand the reason for blocking mcast was because we treated it like broadcast in the past. Now that we have IGMP snooping supported, would it be reasonable to enable mcast by default. (but still keep a config knob to default block it for downstream parity)? The reason I bring this is up is from my basic understanding of k8s the default traffic is "anything can talk to anything" and nothing should be blocked (hence the network policy behavior).
> > ```
> 
> I'm not sure about this, I'll let others comment. The default behavior is to drop multicast but it can definitely be changed if needed.
> 
> > ```
> > 2. Why do we have 2 port groups (1 for deny and 1 for allow) with the current method. Why not just remove pods in annotated namespaces from the original deny port group?
> > ```
> 
> We need 2 port groups. When an ACL is applied to a port group that means that it's applied to all logical switches that have ports in the port group. If two pods P1 and P2 from different namespaces NS1 and NS2 reside on the same node N and multicast is enabled only in NS2, removing P2 from the default deny mcast port group is not enough because the deny multicast ACL will still match on all multicast packets on node N.
> 
> The alternative would be to use allow ACLs that match on inport/outport but that would result way more flows than now.
> 
> > ```
> > 3. Is there any documentation on how this works in our repo? If not, can we please add some?
> > ```
> 
> There's WIP from network policy in general:
> #1182
> 
> > If the answer to number 2 above is, "we only really need 1 port group". Then I would suggest we use that 1 port group for managing multicast, and then create another one with just management ports in it for handling service reject ACL.
> 
> I don't see a problem with having a single port group for multicast deny and service reject ACLs. The requirement for both features is that the port group contains at least one port from all nodes. So why not have a single PG?
> 
> There seems to be some confusion in general about how Port_Group ACLs work so I'm leaving this here for reference:
> 
> ```
>       Access control rules that apply to the port group. Applying an ACL
>       to a port group has the same effect as applying the ACL to all logical
>       lswitches that the ports of the port group belong to.
> ```
> 
> https://github.com/ovn-org/ovn/blob/46b23163a54abb8cef6ae12585985045735aff45/ovn-nb.xml#L1470

@dceara I opened two RFE to address this confusion and to discuss if we can improve Port_Group table to simplify it.
https://bugzilla.redhat.com/show_bug.cgi?id=1885668
https://bugzilla.redhat.com/show_bug.cgi?id=1885670
I will push a patch for BZ#1885668, but we should probably discuss 1885670 and see if we can improve it to come up with better APIs.

**Comment by dcbw** [comment]
> `1. As I understand the reason for blocking mcast was because we treated it like broadcast in the past. Now that we have IGMP snooping supported, would it be reasonable to enable mcast by default. (but still keep a config knob to default block it for downstream parity)? The reason I bring this is up is from my basic understanding of k8s the default traffic is "anything can talk to anything" and nothing should be blocked (hence the network policy behavior).`

Kube NetworkPolicy does not address broadcast/multicast traffic in any way. So NP is not germane to this discussion and does not need to inform our multicast implementation.

You could probably map MC onto the Ingress parts of NP (assuming the source IP is still the source Pod's IP, which may not always be the case). But you can't map it onto Egress parts, since the dest IP is a multicast group.

File: go-controller/pkg/ovn/service.go (hunk 8 above)
Comment (trozet) [reply go-controller/pkg/ovn/service.go:171]: @vishnoianil still missing this?

**Comment by trozet** [comment]
@vishnoianil reviewed your latest version. Looks good besides one comment and unit tests not working. Thanks for reworking the patch into the 2 commits.

File: go-controller/pkg/ovn/service.go (hunk 8 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/service.go:171]: Done.

**Comment by vishnoianil** [comment]
> @vishnoianil reviewed your latest version. Looks good besides one comment and unit tests not working. Thanks for reworking the patch into the 2 commits.

@trozet I fixed the comment and all the checks are passing now. PTAL, thanks

File: go-controller/pkg/ovn/policy_test.go [hunk 10]
```diff
@@ -1198,6 +1189,9 @@ var _ = Describe("OVN NetworkPolicy Operations", func() {
 				_, err = fakeOvn.fakeClient.CoreV1().Namespaces().Update(context.TODO(), ns, metav1.UpdateOptions{})
 				Expect(err).NotTo(HaveOccurred())
 				Eventually(fExec.CalledMatchesExpected).Should(BeTrue(), fExec.ErrorDesc)
+
+				err = fakeOvn.fakeClient.CoreV1().Namespaces().Delete(context.TODO(), namespace1.Name, *metav1.NewDeleteOptions(0))
```
Comment (trozet) [inline go-controller/pkg/ovn/policy_test.go:1193]: why is this added? are we not cleaning up the test correctly?

File: go-controller/pkg/ovn/policy_test.go [hunk 11]
```diff
@@ -1234,25 +1227,29 @@ var _ = Describe("OVN NetworkPolicy Operations", func() {
 						},
 					},
 				)
+				fakeOvn.controller.WatchNamespaces()
 				nPodTest.populateLogicalSwitchCache(fakeOvn)
 				fakeOvn.controller.WatchPods()
-				fakeOvn.controller.WatchNamespaces()
```
Comment (trozet) [inline go-controller/pkg/ovn/policy_test.go:1239]: this indicates there is something else wrong with the patch here. watch namespace handler will handle pods that did not get hit due to the watch ordering for pods being first.

File: go-controller/pkg/ovn/policy_test.go (hunk 10 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/policy_test.go:1193]: no, i think i added to ensure that fakeOvn is really doing the cleanup and not hitting any race condition. Let me remove it.

File: go-controller/pkg/ovn/policy_test.go (hunk 11 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/policy_test.go:1239]: The problem with this ordering is that the test itself misses the namespace events. When i run it locally in this order it passes, but not upstream, so some timing issue here. And seems like this patch is kind of surfacing those issue. Given that i wait for 5 second now before matching the command, ig might work with the ordering, so let me revert and see.

Comment (vishnoianil) [reply go-controller/pkg/ovn/policy_test.go:1239]: But does this ordering matters? because it anyways enable multicast, once pod is part of the namespace, so not sure calling WatchNamespaces before WatchPods changes anything in term of the test?

Comment (vishnoianil) [reply go-controller/pkg/ovn/policy_test.go]: @trozet this is the issue 

```
I1014 20:11:54.568942   14616 reflector.go:175] Starting reflector *v1.EgressIP (0s) from pkg/crd/egressip/v1/apis/informers/externalversions/factory.go:117
I1014 20:11:54.568960   14616 reflector.go:211] Listing and watching *v1.EgressIP from pkg/crd/egressip/v1/apis/informers/externalversions/factory.go:117
I1014 20:11:54.668984   14616 shared_informer.go:253] caches populated
I1014 20:11:54.669298   14616 kube.go:63] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.1.3/24"],"mac_address":"0a:58:0a:80:01:03","gateway_ips":["10.128.1.1"],"ip_address":"10.128.1.3/24","gateway_ip":"10.128.1.1"}}] on pod namespace1/myPod
I1014 20:12:04.670016   14616 pods.go:259] [namespace1/myPod] addLogicalPort took 10.000733062s
**E1014 20:12:04.670049   14616 ovn.go:546] timeout waiting for namespace event <<<<<<<<<<<HERE**
I1014 20:12:04.670083   14616 ovn.go:591] Bootstrapping existing pods and cleaning stale pods took 10.000911062s
I1014 20:12:04.670154   14616 ovn.go:875] Bootstrapping existing namespaces and cleaning stale namespaces took 63.7µs
I1014 20:12:04.670295   14616 pods.go:295] LSP already exists for port: namespace1_myPod
```

File: go-controller/pkg/ovn/policy_test.go [hunk 12]
```diff
@@ -1241,16 +1231,18 @@ var _ = Describe("OVN NetworkPolicy Operations", func() {
 					context.TODO(), namespace1.Name, metav1.GetOptions{})
 				Expect(err).NotTo(HaveOccurred())
 				Expect(ns).NotTo(BeNil())
-				Eventually(fExec.CalledMatchesExpected).Should(BeTrue(), fExec.ErrorDesc)
+
+				ns.Annotations[nsMulticastAnnotation] = "true"
```
Comment (dcbw) [inline go-controller/pkg/ovn/policy_test.go:1235]: @vishnoianil this block that calls Update() needs to be *after* the fakeExec commands are added. Because the Update() could trigger ovn-nbctl calls before the fakeExec stuff is in-place.

Comment (dcbw) [reply go-controller/pkg/ovn/policy_test.go:1235]: Basically, I think you should revert this hunk of the diff

Comment (dcbw) [reply go-controller/pkg/ovn/policy_test.go:1235]: Plus you need to make sure WatchNamespaces() is before WatchPods()

Comment (dcbw) [reply go-controller/pkg/ovn/policy_test.go:1235]: eg, see https://github.com/ovn-org/ovn-kubernetes/pull/1757/files#diff-2a5ea421d13f2f639ad217cada1f5c741e9ba6d68abcf431a700c0f68b71e04dR1224

**Comment by vishnoianil** [comment]
> @vishnoianil WatchNamespaces() should be first, see #1643 to fix that everywhere.

@dcbw Yes, that's what i did and it passed all the CI tests. But @trozet mentioned that we intentionally call watchNamespace later to catch out of order events. So there are bunch of test who misses namespace events (although they pass ).

File: go-controller/pkg/ovn/policy_test.go (hunk 12 above)
Comment (vishnoianil) [reply go-controller/pkg/ovn/policy_test.go:1235]: Yeah, I already tested this. Tests passes if i just move the watchNamespace before watchPods, so i believe moving adding commands before the update() has no impact. But anyways doesn't matter to revert it. @trozet what are your thoughts here? just don't want to go back and forth on this :).

Comment (dcbw) [reply go-controller/pkg/ovn/policy_test.go:1235]: I'm going to strenuously disagree with Tim here; we want https://github.com/ovn-org/ovn-kubernetes/pull/1643 to satisfy both me and Tim, then we always call Namespace before Pod and we're done with it.

**Comment by dcbw** [comment]
> @dcbw Yes, that's what i did and it passed all the CI tests. But @trozet mentioned that we intentionally call watchNamespace later to catch out of order events. So there are bunch of test who misses namespace events (although they pass ).

We do not do that intentionally in the tests, so we should not do that intentionally here. But I have a testcase in https://github.com/ovn-org/ovn-kubernetes/pull/1643 to handle this, which should satisfy Tim.

**Comment by dcbw** [comment]
@vishnoianil can you rebase and squash the 3rd commit (Address review comments) into the first commit (Migrate default deny multicast policy to port-group) (but leave the second commit standalone!) and then I think we're good to go.

**Comment by vishnoianil** [comment]
> @vishnoianil can you rebase and squash the 3rd commit (Address review comments) into the first commit (Migrate default deny multicast policy to port-group) (but leave the second commit standalone!) and then I think we're good to go.

@dcbw Done. PTAL, thanks.

**Comment by dcbw** [comment]
@vishnoianil one more rebase:

```
level=warning msg="[runner] Can't run linter unused: buildir: analysis skipped: errors in package: [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/pods_test.go:464:7: t.addPodDenyMcast undefined (type pod has no field or method addPodDenyMcast)]"
level=info msg="[runner] processing took 4.1µs with stages: skip_dirs: 500ns, nolint: 500ns, max_same_issues: 500ns, cgo: 300ns, max_from_linter: 300ns, identifier_marker: 200ns, autogenerated_exclude: 200ns, path_prettifier: 200ns, filename_unadjuster: 200ns, source_code: 200ns, skip_files: 200ns, max_per_file_from_linter: 100ns, exclude: 100ns, sort_results: 100ns, uniq_by_line: 100ns, path_shortener: 100ns, diff: 100ns, exclude-rules: 100ns, path_prefixer: 100ns, severity-rules: 0s"
level=info msg="[runner] linters took 48.520878546s with stages: goanalysis_metalinter: 43.276113819s, unused: 5.243973614s"
level=error msg="Running error: buildir: analysis skipped: errors in package: [/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/pods_test.go:464:7: t.addPodDenyMcast undefined (type pod has no field or method addPodDenyMcast)]"
```
